File: modules/multiplexer/tmux.go
package multiplexer

import (
	"context"
	"fmt"
	"os"
	"strings"
	"sync"
	"time"
)

// TmuxBackend implements Backend for the tmux terminal multiplexer.
type TmuxBackend struct {
	mu    sync.RWMutex
	panes map[string]*PaneInfo
}

func NewTmuxBackend() *TmuxBackend {
	return &TmuxBackend{
		panes: make(map[string]*PaneInfo),
	}
}

func (t *TmuxBackend) Name() string { return "tmux" }

func (t *TmuxBackend) IsAvailable() bool {
	_, ok := os.LookupEnv("TMUX")
	return ok
}

func (t *TmuxBackend) CreatePane(ctx context.Context, name string, command string) (PaneInfo, error) {
	// -P prints pane ID, -F sets format, -h splits horizontally
	args := []string{"split-window", "-h", "-P", "-F", "#{pane_id}", "-t", name, "sh", "-c", command}
	out, err := runCommand(ctx, "tmux", args...)
	if err != nil {
		return PaneInfo{}, fmt.Errorf("tmux create pane: %w", err)
	}

	paneID := parseTmuxPaneID(out)
	if paneID == "" {
		paneID = fmt.Sprintf("tmux-%s-%d", name, time.Now().UnixNano())
	}

	info := PaneInfo{
		ID:        paneID,
		Name:      name,
		State:     PaneStateRunning,
		CreatedAt: time.Now(),
		Backend:   "tmux",
	}

	t.mu.Lock()
	t.panes[paneID] = &info
	t.mu.Unlock()

	return info, nil
}

func (t *TmuxBackend) ClosePane(ctx context.Context, paneID string) error {
	_, err := runCommand(ctx, "tmux", "kill-pane", "-t", paneID)
	if err != nil {
		return fmt.Errorf("tmux close pane %s: %w", paneID, err)
	}

	t.mu.Lock()
	if p, ok := t.panes[paneID]; ok {
		p.State = PaneStateExited
		delete(t.panes, paneID)
	}
	t.mu.Unlock()

	return nil
}

func (t *TmuxBackend) SendText(ctx context.Context, paneID string, text string) error {
	_, err := runCommand(ctx, "tmux", "send-keys", "-t", paneID, text, "Enter")
	if err != nil {
		return fmt.Errorf("tmux send text to %s: %w", paneID, err)
	}
	return nil
}

func (t *TmuxBackend) FocusPane(ctx context.Context, paneID string) error {
	_, err := runCommand(ctx, "tmux", "select-pane", "-t", paneID)
	if err != nil {
		return fmt.Errorf("tmux focus pane %s: %w", paneID, err)
	}
	return nil
}

func (t *TmuxBackend) CaptureOutput(ctx context.Context, paneID string) (string, error) {
	out, err := runCommand(ctx, "tmux", "capture-pane", "-t", paneID, "-p")
	if err != nil {
		return "", fmt.Errorf("tmux capture output %s: %w", paneID, err)
	}
	return out, nil
}

func (t *TmuxBackend) ListPanes(ctx context.Context) ([]PaneInfo, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()

	panes := make([]PaneInfo, 0, len(t.panes))
	for _, p := range t.panes {
		panes = append(panes, *p)
	}
	return panes, nil
}

// parseTmuxPaneID extracts the pane ID from tmux output (e.g., "%5").
func parseTmuxPaneID(output string) string {
	output = strings.TrimSpace(output)
	if output == "" {
		return ""
	}
	// tmux pane IDs start with %
	for _, line := range strings.Split(output, "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "%") {
			return line
		}
	}
	return output
}

File: packages/multiplexer/src/tmux.ts
import { PaneState, type PaneInfo } from "./pane.js";
import type { Backend } from "./backend.js";
import { runCommand } from "./backend.js";

/**
 * TmuxBackend implements Backend for the tmux terminal multiplexer.
 */
export class TmuxBackend implements Backend {
  private panes = new Map<string, PaneInfo>();

  /** name returns the backend identifier used for backend selection. */
  name(): string {
    return "tmux";
  }

  /** isAvailable reports whether we are inside a tmux session (TMUX env set). */
  isAvailable(): boolean {
    return !!process.env.TMUX;
  }

  /**
   * createPane splits the current tmux window horizontally and runs the command.
   * Falls back to a synthesized pane ID if tmux output parsing fails.
   */
  async createPane(name: string, command: string): Promise<PaneInfo> {
    // -P prints pane ID, -F sets format, -h splits horizontally
    const args = ["split-window", "-h", "-P", "-F", "#{pane_id}", "-t", name, "sh", "-c", command];
    const out = await runCommand("tmux", args);

    let paneID = parseTmuxPaneID(out);
    if (!paneID) {
      paneID = `tmux-${name}-${Date.now()}`;
    }

    const info: PaneInfo = {
      id: paneID,
      name,
      state: PaneState.Running,
      createdAt: new Date(),
      backend: "tmux",
    };

    this.panes.set(paneID, info);
    return info;
  }

  /** closePane kills the named tmux pane and drops it from the local registry. */
  async closePane(paneID: string): Promise<void> {
    await runCommand("tmux", ["kill-pane", "-t", paneID]);
    const p = this.panes.get(paneID);
    if (p) {
      this.panes.delete(paneID);
    }
  }

  /** sendText sends a line of text into the pane followed by Enter via send-keys. */
  async sendText(paneID: string, text: string): Promise<void> {
    await runCommand("tmux", ["send-keys", "-t", paneID, text, "Enter"]);
  }

  /** focusPane brings the pane to the foreground via select-pane. */
  async focusPane(paneID: string): Promise<void> {
    await runCommand("tmux", ["select-pane", "-t", paneID]);
  }

  /** captureOutput returns the visible pane content via capture-pane -p. */
  async captureOutput(paneID: string): Promise<string> {
    return runCommand("tmux", ["capture-pane", "-t", paneID, "-p"]);
  }

  /** listPanes returns the locally-tracked panes created by this backend instance. */
  async listPanes(): Promise<PaneInfo[]> {
    return [...this.panes.values()];
  }
}

/**
 * parseTmuxPaneID extracts the pane ID from tmux output (e.g., "%5").
 */
function parseTmuxPaneID(output: string): string {
  const trimmed = output.trim();
  if (!trimmed) return "";

  for (const line of trimmed.split("\n")) {
    const l = line.trim();
    if (l.startsWith("%")) {
      return l;
    }
  }
  return trimmed;
}
